### ISSUE TITLE
[Parallel rendering] Masking

### ISSUE THREAD

**Issue by kisvegabor** [kisvegabor/lvgl#4059]
Part of #4016

### Introduce the problem
In the current stage of parallel rendering an lot of drawing related features are disabled. We should reconsider and re-implement them. This issue is to discuss how to handle masks, effects, etc in the new parallel render architecture.

### Examples and cases

Here are a few use-case for which we should have a solution in the end. I also add my proposed solution (see their description in the next section)

**Mixed use cases**
- Alpha gradient: Use `lv_color32_t` in APIs and make gradients deal with the alpha channel too.
- Music player demo stripes around the cover: Use triangles (now they are trapezoids but in v9 Sw render shouldn't support polygons)
- Chart faded area (like in the widgets demo): Use right angle triangles with gradient
- Transform widget parts (e.g. slider knob rotation) : Allow easily creating and managing new layers manually  in draw events
- Masking
   - Clip corner: Built-in solution for better optimization, render children to a new layer if they overlap the corner and apply mask
   - Bitmap or text mask to any drawing or layer: There are special "mask draw tasks". If they are added the widget are drawn to a new ARGB layer and that layer is masked by the masking draw tasks.
- Blur and other filters: `post process cb`
- Backdrop filter (e.g. blur): ???
- Drop shadow on ARGB images: `post process cb`
- Real time outline for ARGB images: `post process cb`
- GPU effects (e.g. contrast): `post process cb`
- Perspective transform: just implement it
- SVG/vector graphics: generic API  + 3rd party lib 

**Software rendered primitives** 
- Line:
    - idea: use A8 image rotation 
    - dash skewed: dash the A8 image
    - square and rounded line ending: add rounded caps to the A8 image
    - line joint: not supported, use SVG instead  
- Arc
  - square and rounded line ending: custom mask, some local masking magic 
  - process in 16x16 blocks to easily ignore chunks in the middle 
- Rounded rectangle and border
  - different radius on corners
  - process large radius in 16x16 chunks to have simplify squares on not radius parts.
- Triangle
   - simple anti-aliasing
   - polygons are not supported: use SVG lib if needed 
- Image
  - Transform the image in 8x8 squares to easily discard 64 px if they are out of the original image

**Not strictly related**
- Text on path
- Text shadow
- Gradient on border and shadow
- Multiple border,s outlines and shadows

### Suggested solution
- Have built-in mask-related draw tasks applied only on ARGB images (layers). It's like `LV_DRAW_TASK_TYPE_IMAGE` but it performs masking. The following mask sources should be supported:  bitmap,  text, rounded rectangle
- `post_process cb`:
  - A style property which can add new draw tasks e.g. for masking, blurring, etc. It's called when all children's draw tasks are added. If there is a `post_process cb` the widget with its children are drawn on a new layer and that layer can be manipulated independently of the other parts of the screen.  
  - An other option is to add new draw tasks in `LV_EVENT_DRAW_POST_END`. The object needs to know to create a new layer. 
- clip corner:  A special mask (i.e. rectangle mask with radius). It should be applied only on the direct children if they are on a radius area. Not called from `post_process_cb` as LVGL needs to do some special considerations with it and enable it only if needed.
- new layer during drawing: Make it simple to manually create nee layers. E.g. in a draw begin event create a layer and continue the drawing there. In a draw end event merge the layer back to the original (parent) layer. It can be done automatically if a part has transformations, or anything requiring a new layer.

---

I'm thinking about these for months and I believe it's hard to see how the pieces fit together, especially as it's heavily related to the new parallel rendering  architecture. However, if you have any idea or remark about, I'd be more than happy the hear it :slightly_smiling_face: 

I can take care most of the implementation but it'd be amazing if you could bring up other use cases that we should consider.

**Comment by GorgonMeducer** [comment]
> Alpha gradient: Use lv_color32_t in APIs and make gradients deal with the alpha channel too.

Do you plan to use ARGB8888 for internal processing?

**Comment by kisvegabor** [comment]
> Do you plan to use ARGB8888 for internal processing?

Definitely not. Just in the API to set the colors.   E.g. before filling an area with a color the ARGB8888 color should be converted to e.g. RGB565 (according to `LV_COLOR_DEPTH`).

**Comment by GorgonMeducer** [comment]
If that's the case, I fully agree with you that using lv_color32_t in API, and is definitely a good idea. 
About gradient, I notice that LVGL only supports two gradient modes, horizontal and vertical. I would suggest you adding support to points based gradient, for example:

- generate gradient based on two points (backward compatible with horizontal and vertical mode when either x is the same or y is the same)
- generate gradient based on three points
- generate gradient based on four corner points

By the way, to avoid IP pollution, as I am implementing the points-based gradient in arm-2d, I am not able to contribute this feature, sorry. 

For ARGB8565, I highly recommend deprecating this color format and moving to RGB565A8.

**Comment by kisvegabor** [comment]
> About gradient, I notice that LVGL only supports two gradient modes, horizontal and vertical. I would suggest you adding support to points based gradient, for example:

You can set the number of max gradients point with [`LV_GRADIENT_MAX_STOPS`]( https://github.com/lvgl/lvgl/blob/master/lv_conf_template.h#L317).

In a minor version I agree to add more gradients, like skew, circle or conic.

> For ARGB8565, I highly recommend deprecating this color format and moving to RGB565A8.

We are using it internally when a new layer is created and it needs to have alpha channel too. I'll check if we can switch to `RGB565A8`. I'm concerned that it will cause too many branches due to the not tightly packed pixels.

**Comment by nicusorcitu** [comment]
The new VGLite GPU will support more features than RT1170 - almost everything is needed for drawing:
**- It will support A8 masking as well as path masking**
- It will support gaussian blur - so we can draw also the shadows
- draw letters (path letters and raster letters)

I believe we shall target the LVGL re-architecture not only for simpler blitters (like PXP) but also for complex GPUs (like VGLite). Maybe at some point the GPU will take care of all of the draw request. Just splitting them in very small parts and queuing them to GPU (even as a raw) will not be efficient.

**Comment by kisvegabor** [comment]
It sounds amazing! It seems to me these features can be realized with the planned architecture as we describe the draw tasks from a quite high level.

The only thing is not clear to me is how to handle RGB565 + alpha. Now LVGL uses the non standard ARGB9565 color format but @GorgonMeducer [suggested](https://github.com/lvgl/lvgl/issues/4059#issuecomment-1468230452) using RGB565+A8 (all colors first and all alpha bytes after that). I think RGB565+A8 would be easier to handle with VG-Lite too. What to do you think?

**Comment by tdjastrzebski** [comment]
I think this discussion ought to be put into broader perspective.  
What is the current goal of LVGL?  What is the ambition and motivation?  
Is it the best performance on current mainstream hardware or, maybe, features are even more important than performance?  
If so, what hardware is currently the most popular or what features are missing the most? What new hardware is coming?  
Or, maybe what is needed the most is stability and quality?
What improvements are needed the most and how do you know that?
I think it is important to know the answers to make sure architecture, including color format, supports the current goal(s).

**Comment by kisvegabor** [comment]
Great questions!

> What is the current goal of LVGL? What is the ambition and motivation?

I couldn't say a  single stunning marketing sentence but I see these:
- remain the most popular free and open source embedded GUI library
- keep the platform independent and easy to integrate nature of the library
- keep its easy to use nature
- while supporting low end devices and have low memory footprint, support high-end devices too (powerful GPUs, video cards, multi core MCUs, etc)
- have features allowing to create stunning UIs
- encourage contribution
- reach financial  stability, i.e. find some funds, ideally have some full time employees too (I'm already on it) 


> Is it the best performance on current mainstream hardware or, maybe, features are even more important than performance?

As the performance of the hardware is growing rapidly, on the long run features will be more important. However, the low end devices will be with us too for many years and we should support them too. So we should have highly modular ans scalable architecture.

> If so, what hardware is currently the most popular or what features are missing the most? What new hardware is coming?
Or, maybe what is needed the most is stability and quality? What improvements are needed the most and how do you know that?

I think many things are already working well: we have many widgets, great style and layout systems, multi language support, low memory footprint. However as I see  two main things are missing and we try to address these in v9:
- have drivers in LVGL to make it easier to get started
- more effective support for MCU GPUs and 3D GPUs too. 


I'd be very curious about your opinion too. Maybe we should open a new off-topic issue for it as this is not strictly related to how we should implement masking.

**Comment by tdjastrzebski** [comment]
Let's take goals 1 and 5, because others kind of only support them:
* **have features allowing to create stunning UIs**
* **remain the most popular**  

Now, among whom? Enthusiasts or professionals?
If financial stability is considered, professionals may count more, but it is only my blind assumption and could be wrong.  
But let's assume I am correct. What do professionals want? What hardware do they use these days designing `new` solutions?

E.g. how would you go about creating real, marketable packet mp3 player with a stunning UIs? I mean, on embedded device, because DirectX, Unreal Engine etc. or even Chromium is not something to try to compete against. What hardware would you choose? - especially knowing that GUI, more often than not, although important, is just a GUI and device must (e.g.) primarily smoothly play mp3, with no hiccups when rendering screen? Do you see my point?  
If so, what embedded graphics accelerators support is the most critical today?

**Comment by tdjastrzebski** [comment]
Today my hardware pick would probably be STM32U5 - see [this video I just made](https://youtu.be/WbMabPVxtAw).
Surely NXP VGLite GPU is another one. What are the other best options?

**Comment by GorgonMeducer** [comment]
STM32U5 with DMA2D is an appropriate choice as a mainstream solution. It is based on Cortex-M33 (slightly better than Cortex-M4F in terms of performance). LVGL should deliver 30FPS performance for most of the scenarios. 

If we target M7-based devices (and VGLite style of GPU) and adjust our expectations for resource and performance, the mainstream devices, i.e. M4/M33-based devices, will suffer. 
The total volume of the M33/M4-based products is much higher than M7-based products. 

My suggestions about the LVGL:

- M4/M33+DMA2D-like acceleration: 30FPS for most of the scenario
- M7 + 2D-GPU: 60FPS for most of the scenario

**Comment by kisvegabor** [comment]
I've just opened https://github.com/lvgl/lvgl/issues/4073 and summarized my thought in more structured way. Could you comment on it?

**Comment by tdjastrzebski** [comment]
Exactly. New STM32U5xx does not only have DMA2D, which good support, I believe, should be among TOP priorities.  
STM32U5 also comes with NeoChrom GPU aka ***GPU2D***. I think currently it is only supported by TouchGFX. This is what makes  [this graphics](https://youtu.be/WbMabPVxtAw) so ass-kicking. Part of it is SVG - hardware supported!
Besides DMA2D, this would be my next target because this is where embedded technology seems to be going - less powerful core, but way more powerful graphics. I think this is very relevant to this thread because whatever choices are made, they should support the goal(s).

**Comment by nicusorcitu** [comment]
> The only thing is not clear to me is how to handle RGB565 + alpha. Now LVGL uses the non standard ARGB9565 color format but @GorgonMeducer https://github.com/lvgl/lvgl/issues/4059#issuecomment-1468230452 using RGB565+A8 (all colors first and all alpha bytes after that). I think RGB565+A8 would be easier to handle with VG-Lite too. What to do you think?
>
The old VGlite GPUs does not have support for neither A8RGB565 or RGB565A8. They will not support masking as was till present.

But the new VGLite will have support for both, not sure which one will be more convenient. Probably is the same performance.

> - It will support A8 masking as well as path masking
>
What I am trying to say is the new VGLite GPU will support A8 masking, which is different the way LVGL approach it.
My understanding is that you desire to apply masking at the end of drawing, e.g. if you want to draw this
![graph](https://user-images.githubusercontent.com/97085059/226322824-ce5a3e41-6fd3-4410-9701-9096b3f42579.png)
you will draw a full rectangle with gradient, then you will apply (blend the result with) an A8 mask so that the left up corner will become transparent. Am I right?
With A8 VGLite masking, you do not have to do that in two steps. And you do not have to draw the full rectangle.
You can pass the A8 mask to the GPU so that it will draw the non masked pixels (the visible part of the gradient). So more efficient.

Having that, is it possible to take into consideration to keep the rectangle task with gradient and mask? At least for the beginning until we find out more possibilities on how to approach masks in general.

**Comment by GorgonMeducer** [comment]
> But the new VGLite will have support for both, not sure which one will be more convenient. 

@nicusorcitu ARGB8565 have 3 bytes per pixel, which leads to unaligned access on the bus, RGB565A8 has a separate colour array and an A8 array, so there always be aligned access on the bus. That's the difference in terms of performance. 

I am not sure how VGLite fetches data, but for the C algorithm, ARGB8565 isn't as efficient as RGB565A8. 

> With A8 VGLite masking, you do not have to do that in two steps. And you do not have to draw the full rectangle.
You can pass the A8 mask to the GPU so that it will draw only the visibile part of the gradient. So more efficient.

@kisvegabor FYI, arm-2d uses the same method, it is called filling-colour-with-mask-and-opacity. You can consider to update the render and take advantage of those features.

**Comment by kisvegabor** [comment]
> FYI, arm-2d uses the same method, it is called filling-colour-with-mask-and-opacity. You can consider to update the render and take advantage of those features.

Besides filling with a color what other operation can VGLite and arm-2d do with A8 masks? If it's only filling we can have a special draw task for it.

Note that, I wouldn't like to force any GPU to approach rendering as the SW render does. (In the SW render we have a lot of limitations and trade offs.) Instead we should have high level and abstract draw tasks that can be drawn as the GPU wishes.

E.g.  this special chart section is now drawn by masks which is very specific. 
![image](https://user-images.githubusercontent.com/7599318/226375690-e92f8a98-da8d-4c23-899d-d04cfc817281.png)

Instead these should be right angled triangles with a gradient. A triangle is much more generic and it's up to the GPU how to draw it.

Another option is to make LVGL (SW render and GPUs) capable of rendering in A8 too (besides the selected native color depth). This way LVGL could render masks and could say to use these masks for filling. However it'd have some disadvantages:
-larger code size
-the mask needs RAM 
-it's less generic as we say that the chart's triangle should be drawn with a mask, but maybe the GPU could do it faster/better

So in general I'd say having generic primitives seems reasonable. If we need special masks or effects for any reason (clip corner, fading all children, blur, etc), LVGL will create a layer (image) and do the operations on it.

**Comment by GorgonMeducer** [comment]
> Besides filling with a color what other operation can VGLite and arm-2d do with A8 masks? If it's only filling we can have a special draw task for it.

For arm-2d, we have:
- copy-with-mask-and-opacity (already used in accelerating LVGL). 
- tile-fill-with-mask-and-opacity
- filling-colour-with-mask-opacity-and-transform

**Comment by kisvegabor** [comment]
Thank you. 

I'm thinking about if it's worth to add an A8 mask parameter to all primitive drawings but I can not really find a use case when a simple draw task should have a bitmap mask. E.g. to render a rounded rectangle arm-2d can get a rounded rectangle mask from the SW render's mask engine. But VG-Lite can just draw a rounded rectangle on its own without any masks.

But what would be use case when the bitmap mask is already known on draw task level. Or just leave this opportunity open for any custom drawing?

**Comment by tdjastrzebski** [comment]
Form ST DMA2D perspective I think it would be the best to internally always use ARGB8888 interim buffer.
Although in case of 16 color DMA2D can blend A8 with RGB656 and produce ARGB8888, it cannot combine bitmap with a mask and blend with a background in one operation.
It needs to do it in two steps using interim buffer, which every time has to be created and then released unless one permanent buffer is kept which could only be increased if/when needed but then more total memory is required and two blending operations are needed while when always using ARGB8888 buffer no extra memory and mixing would be necessary.
What I suggest is instead of using ARGB565 and A8 mask always use ARGB8888 for both 16 and 32b color.
32b color and RGB888+A8 mask is slightly less problematic as long as LVGL actually uses 32b xRGB8888 buffer and A8 can be simply copied over (mixed) without a need of creating an extra buffer for that, albeit on Cortex-M7 every time SW mixing is used L1 buffer cache area needs to be cleaned before DMA2D can fetch it without glitches, which adds an overhead.

**Comment by GorgonMeducer** [comment]
at least, @tdjastrzebski and I don't like ARGB8565. 

By the way, Cortex-M0/0+, Cortex-M3, Cortex-M4F and Cortex-M33 processors are using AHB-Lite which is a 32bit bus. Using RGBA8888 as internal color format isn't as efficient as RGB565 + A8 in terms of bus bandwidth.

Meanwhile the Cortex-M7, Cortex-M55 and Cortex-M85 are using 64bit AXI bus interface, hence ARGB8888 can take advantage of the 64bit bandwidth. 

So, in mid term, as we take Cortex-M33 as the mainstream platform, it is better not using ARGB8888 as the internal processing format. In long term when Cortex-M55 becomes the mainstream, it is good to move to ARGB8888. I guess that would happen when we are developing LVGL11.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer I do not think I entirely agree, not without hard performance data. AHB is not the only factor and I believe other reasons to use ARGB888 may still prevail.

@kisvegabor Btw, maybe it would be a good idea to merge PR #3904 (stm32 dma2d rework) with rev 9? I mean, create a new one.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer Besides, I think that the decision which platform currently (e.g. M4/M7/M33/M55) should have the most attention should be based on **strong evidence**. By no means I am saying that I disagree, but how do we actually know?

@kisvegabor I guess, what I really suggest is implementing more formal Product Management practices. In my view, LVGL reached the stage where it is a must to successfully move forward.

**Comment by nicusorcitu** [comment]
> Besides filling with a color what other operation can VGLite and arm-2d do with A8 masks?
>
The mask enablement is global, once it it set it will be valid for all kind of operations blit/draw until is disabled.

![226375690-e92f8a98-da8d-4c23-899d-d04cfc817281](https://user-images.githubusercontent.com/97085059/226567136-09452e94-6362-4e5d-af92-9ada5a34fe5b.png)

> E.g. this special chart section is now drawn by masks which is very specific.
> Instead these should be right angled triangles with a gradient. A triangle is much more generic and it's up to the GPU how to draw it.
> 
For this specific case, it is fine if we draw a rectangle and a rectangle with gradient. Only that you have to recalculate the gradient for triangle. But should be fine even thought it can be done as a single time.

> I'm thinking about if it's worth to add an A8 mask parameter to all primitive drawings but I can not really find a use case when a simple draw task should have a bitmap mask. E.g. to render a rounded rectangle arm-2d can get a rounded rectangle mask from the SW render's mask engine. But VG-Lite can just draw a rounded rectangle on its own without any masks.

I can give you two more examples of masking:
1. the meeter
![Screenshot 2023-03-21 132316](https://user-images.githubusercontent.com/97085059/226592141-07248223-e69b-425c-8e52-e5a970eeb5cd.png)
All the lines from are draw by having two points (one is the center of the circe and the other one on the circle circumferince). These lines are "shorted" because in the middle of the meter there is a mask "cutting" those lines. Without the mask the lines would have been till the center of the circle.
With next VGLite for example we can set that mask once and then draw every line till the end when we can disable the mask. We don't have to draw the full line and then blend with A8.

2. slide bar
![Screenshot 2023-03-21 132651](https://user-images.githubusercontent.com/97085059/226592912-ce2745b7-e72c-40a5-941c-21c4a67d97b0.png)
Without the mask it would have looked like this:

![Screenshot 2023-03-21 132729](https://user-images.githubusercontent.com/97085059/226593073-3e3c1b98-985f-4c21-9b15-04940902aaf0.png)

With VGLite we can also rotate widgets. We can draw lines, rectangles, etc and rotate them at any degree.

**Comment by kisvegabor** [comment]
> Form ST DMA2D perspective I think it would be the best to internally always use ARGB8888 interim buffer.
Although in case of 16 color DMA2D can blend A8 with RGB656 and produce ARGB8888, it cannot combine bitmap with a mask and blend with a background in one operation.

We should also consider memory usage and memory bandwidth.  

Anyway, if the user wants they still can use ARGB8888 regardless to the color depth of LVGL. But there should be a draw unit which can draw the ARGB8888 image on a RGB565 or ARGB8565 (or A8RGB565) buffer. If the DMA2D draw unit will support it, it can take this draw task and handle it. If the user uses an ARGB8565 image, and DMA2D cannot support it, it can leave it to SW render and pick another draw task for itself.

> Btw, maybe it would be a good idea to merge PR https://github.com/lvgl/lvgl/pull/3904 (stm32 dma2d rework) with rev 9? I mean, create a new one.

In v9 will will have `draw_unit`s which is a different concept that we have now. The current DMA2D can be mostly reused, but won't be compatible with v9 as it is. So I'd merge it when the new drawing architecture is added too.

> I guess, what I really suggest is implementing more formal Product Management practices. In my view, LVGL reached the stage where it is a must to successfully move forward.

I absolutely agree. We need to step one level else LVGL will collapse at some point. It includes establishing a financial background,  create a clear decision making procedure (with selected decision makers), find responsible people for each topic (architecture, drivers, docs, examples, etc). As I mentioned in the other issue I started to talk about it with vendors, and I hope we can reach to an agreement. It will help to have some founds, and some employees too. Which would be the most important at this point in order to scale the project.

@nicusorcitu 

> These lines are "shorted" because in the middle of the meter there is a mask "cutting" those lines. Without the mask the lines would have been till the center of the circle.

It's just a hack because the 1 pixel resolution of line drawing wasn't good enough and the lines didn't look even. In v9 we can have make the `draw_line` function to accept scaled up coordinates with e.g. 1/16 pixel resolutoin .

> slide bar
Without the mask it would have looked like this:

I was also thinking about it and with some extra calculations in `lv_slider` we can eliminate the use of masks here.

Ideally (and it seems doable) the widgets should not use masks directly. Some masking magic should be used only if the user wants to create something really custom.

> With VGLite we can also rotate widgets. We can draw lines, rectangles, etc and rotate them at any degree.

It's interesting. So maybe we should add an angle + scale parameter to all primitive drawing? I was planning to support it via layers, and the layer as an image is rotated with all the children. Drawing just a rotated rectangle might be useful in some cases too, but I'm not sure it's common enough to be worth the complication. What do you think?

**Comment by nicusorcitu** [comment]
> So maybe we should add an angle + scale parameter to all primitive drawing? I was planning to support it via layers, and the layer as an image is rotated with all the children. Drawing just a rotated rectangle might be useful in some cases too, but I'm not sure it's common enough to be worth the complication
>
Speaking from VGLite perspective, we can have angle/scale to all draw tasks. But then even the CPU itself will have to draw the same (because VGLite and CPU will be two PEs drawing the same widget). So indeed it will complicate a bit the things.
For now we can leave as it is but we can think in the future to add transformation to all primitive drawing.

**Comment by tdjastrzebski** [comment]
@kisvegabor Always using internal 32b buffer (ARGB8888/xRGB8888+A8) is just a thought, but maybe not such a crazy one.
I think it would simplify many things. The real question is whether or not the **target technology** is already there to handle it without significant performance impact. If not yet, it will be there - it is just a question of time, but how do we know when this time arrives? How will we know that it no longer makes sense to spend effort on maintaining both 16 and 32b graphic? What is the **threshold**? Converting from ARGB8888 to RGB565 in `flush_cb()` is a simple task if/when needed, with or without DMA2D so  16b displays can always be supported.

**Comment by GorgonMeducer** [comment]
> Besides, I think that the decision which platform currently (e.g. M4/M7/M33/M55) should have the most attention should be based on strong evidence. By no means I am saying that I disagree, but how do we actually know?

You can start a survey and find the platform with the highest volume. Or you can check with silicon vendors, ST for example, why they keep adding 2D GPU to the mainstream Cortex-M processors while DMA2D is already there.

I cannot provide you with strong evidence. I cannot. Sorry about that.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer How about a survey among LVGL users? What projects do they start, what MPUs do they use/consider?
>  Using RGBA8888 as internal color format isn't as efficient as RGB565 + A8 in terms of bus bandwidth.

I do not disagree, but how much less efficient? That's what I meant by hard data.

**Comment by GorgonMeducer** [comment]
> How about a survey among LVGL users? What projects do they start, what MPUs do they use/consider consider?

It's good idea, and we need the support from @kisvegabor .

> I do not disagree, but how much less efficient? That's what I meant by hard data.

Just run the benchmark (set depth to 16bit and 32bit) and compare the result of topics that involve ARGB.

The interesting part is:

- Suppose RGBA8888 is less efficient than RGB565-A8 but the difference is negligible, then memory footprint is important. 
- Suppose RGBA8888 is efficient than RGB565-A8 but the difference is small, then memory footprint is important
- Suppose RGBA8888 is more efficient than RGB565-A8, then we can suggest users use ARGB8888 as the internal processing format and convert it to RGB565 before flushing as long as the memory footprint is not important. 
- Suppose RGBA8888 is less efficient than RGB565-A8 and the difference is obvious, then using RGB565-A8 is the optimal solution in both performance and memory footprint perspectives. 

See the interesting part here? memory footprint plays an important role, and hardly any one can ignore it in mainstream processors. The question is, why a product select a mainstream processor rather than M7/M85?

The answer is cost. And hence memory footprint is important. Because, as long as there are any RAM or Flash left, new features of the product can be added to make the product more attractive without increase the bom cost. Or people would like to leave some room for future upgrade without changing the bom cost.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer I would add one more yet important factor: CPU usage. DMA2D and similar 'accelerators', if used efficiently, offload CPU. It may make a lot of sense to use it even if FPS is actually lower, but much more cycles can be used for stuff other than graphic in exchange. Let's do not forget that nice graphics is often only 'nice to have', but it is NOT a top feature, high FPS is not a top requirement or benchmark as long as it is 'good enough'. That is why I advocate for LVGL architecture which at least enables efficient DMA2D usage. GPU2D support is next, but it may not be such a 'low hanging fruit'.

**Comment by GorgonMeducer** [comment]
@tdjastrzebski I agree with you about the "good enough" statement. Based on this, reducing the memory footprint is much more reasonable because If we are happy with good enough performance, then people will care more about the memory footprint., especially sparing more of the SRAM used in "good enough" graphics and using them in applications. 

I agree that using a hardware accelerator to offload the CPU is important. It is good to have a DMA2D that uses RGB565 plus A8 as input and output.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer One of the troubles I discovered reimplementing ST DMA2D support (#3904) was that blending RGB565+A8 **with background** requires additional interim memory buffer, which is not the case when xRGB8888+A8 is used. The reason is that neither RGB565+A8 nor xRGB8888+A8 can be used by DMA2D directly but xRGB8888+A8 is easier to handle, because it is easier to produce ARGB8888 from xRGB8888+A8 than from RGB565+A8 before DMA2D can take it over. Note that DMA2D does not support 24b ARGB8565 so RGB565+A8 must be converted to ARGB8888 first and that is where extra RAM is required or DMA2D cannot be used in such case - and currently it is not.

**Comment by kisvegabor** [comment]
@tdjastrzebski 

I seems to me the main motivation behind suggesting to use ARGB8888 internally is to allow DMA2D (and similar GPUs) work more effectively.

I believe the  architecture of v9 is flexible enough to handle this. Assume these cases
- The display has normal RGB565 format but the user want to use an RGB image: No problem DMA2D also knows RGB565, so RGB565 images can be used directly.
- The display has normal RGB565 format but the user wants to use ARGB image. 
    - Option 1: Use ARGB8565 or RGB565 + A8 image. DMA2D cannot be used
    - Option 2 : Use ARGB8888 images and let DMA2D do the pixel format conversation
- The display has RGB565 + alpha format: Not typical as hardware doesn't support ARGB8565 and ARGB8888 is used instead 
 - An internal LVGL  layer has ARGB8565 format: It can really happen e.g. when a rectangle with rounded corners are rotated. DMA2D can't be used for this format at all. Unless, we allow controlling the color format of the layers too. So the layer's color format could be set to ARGB8888 if we want (based on some policies). The problem with that is  `LV_COLOR_DEPTH` is `16` but we will want the SW render to work in 32 bit mode. So we need use different color depths at run time. I wanted to avoid it for the sake of simplicity and code size but let's discuss it and we will see.
  
> Note that DMA2D does not support 24b ARGB8565 so RGB565+A8 must be converted to ARGB8888 first and that is where extra RAM is required or DMA2D cannot be used in such case - and currently it is not.

It's very important that the images can have any format. The the only thing matter is that in the end they need to rendered in a the color format related to `LV_COLOR_DEPTH`. So it's absolutely possible to have an ARGB8888 image and render it to an RGB565 buffer by letting DMA2D make the color conversation on the fly.

**Comment by GorgonMeducer** [comment]
> The display has RGB565 + alpha format: Not typical as hardware doesn't support ARGB8565 and ARGB8888 is used instead

The RGB565-A8 is just for storing PNG pictures, and no need for a display to support RGB565-A8. A8 there is only for alpha-blending.  So RGB565-A8 is used for normal RGB565 LCDs, just like ARGB8565. Or you can say, there is no LCD support ARGB8565. 

Both ARGB8565 and RGB565-A8 are designed to store PNG pictures, and the target LCD uses RGB565.

**Comment by kisvegabor** [comment]
> Both ARGB8565 and RGB565-A8 are designed to store PNG pictures, and the target LCD uses RGB565.

See 
> An internal LVGL layer has ARGB8565 format: It can really happen e.g. when a rectangle with rounded corners are rotated. 

In this case LVGL creates a snapshot about the widget and all its children and rotates it as an image. But if there are holes on the widget (e.g. rounded corner) the widget must be rendered as ARGB. If `LV_COLOR_DEPTH` is `16` then it means ARGB8565.
  
So RGB565 + A8 is really an option as target rendering color format too.

**Comment by GorgonMeducer** [comment]
> RGB565 + A8 is really an option as target rendering colour format too.

OK, I didn't notice this before. In terms of memory usage, it is the same as ARGB8565, and it is more efficient (because of fully aligned memory access). I suggest using RGB565-A5 most of the time.

ARGB8565 is not Vector friendly, so it is tough to accelerate ARGB8565 pixel processing using a SIMD instruction set, which means no DSP,  NEON or Helium. 
RGB565-A8 is Vector friendly so that you can accelerate the pixel processing with DSP, NEON and Helium...

**Comment by nicusorcitu** [comment]
> In terms of memory usage, RGB565-A8 it is the same as ARGB8565, and it is more efficient (because of fully aligned memory access).
>

The reason you suggest is more efficient is that ARGB8565 is packed as 3 bytes, while RGB565 is packed 2 bytes and A8 is packed separated as1byte.

When the GPU needs to fetch a pixel, it will fetch twice for RGB565 A8. It needs to read from two different addresses.

Instead, when GPU fetch a pixel from format ARGB8565, it will read only once. And the remaining 1 byte will be cached for the next operation. Am I right?

How is the  RGB565 A8 more efficient?

**Comment by GorgonMeducer** [comment]
@nicusorcitu You are talking about THE GPU... and only about the THE GPU you are talking.

I can raise a different example: 

other GPU or 2D-capable DMA might have a different understanding. Just like the DMA-350 can take RGB565 array, background pixels and A8 array with different channels and compose them together with a stream interface (so you can plug an alpha-blending logic to that stream interface) and store the result to a output channel. 

It is more efficient. 

If we go back to software processing, for those MCU without DMA2D, RGB565-A8 is more efficient than ARGB8565. 

GPU is a dedicated hardware logic, which means that as long as there are applications and scenarios, people can customize hardware logic and create something new. For existing GPUs, yes, you are right, but new 2D-GPUs are emerging. 

And I have explained, ARGB8565 isn't vector friendly.

**Comment by tdjastrzebski** [comment]
@kisvegabor 
> the main motivation behind suggesting to use ARGB8888 internally is to allow DMA2D (and similar GPUs) work more effectively.

correct

ST DMA2D can handle RGB565/RGB888 conversions back and forth, with or without alpha. That is not an issue.
The problem occurs only when blending of RGB565+A8 image **with existing background** is required. In such case RGB565+A8 image+mask needs to be converted to ARGB8888 first. The same is true for xRGB888+A8, but RGB565+A8 requires an extra buffer while xRGB8888+A8 does not since xRGB8888 has one unused (x) byte where A8 can be simply copied. Of course, this creates some CPU overhead or additional DMA2D transfer, but at least no extra RAM is required.

That is why I suggest always using ARGB8888 instead of RGB565+A8, but only for the relatively small internal buffer which keeps image to be blended, not for anything else like flash-stored graphics etc.
Would that be justifiable? I do not now.
I think memory-wise we are looking at total 25% more RAM for 16b masked image vs creating additional temp 32b buffer at least for the part of this image when it is time to blend it with background.
16b images without a mask are not an issue.

**Comment by GorgonMeducer** [comment]
> always use ARGB8888 instead of RGB565+A8, but only for the relatively small internal buffer, which keeps images to be blended, not for anything else like flash-stored graphics etc.

You are right. We can add a macro option for this. 
For those devices that have no DMA2D, we can disable this feature. And for Cortex-M55/M85, which can accelerate RGB565-A8 using helium, I also suggest disabling this feature.

**Comment by kisvegabor** [comment]
@tdjastrzebski 
> The problem occurs only when blending of RGB565+A8 image with existing background is required. In such case RGB565+A8 image+mask needs to be converted to ARGB8888 first.

However the user always have the opportunity to use ARGB8888 images even if `LV_COLOR_DEPTH` is `16`. We just need to make it clear using ARGB8888 image need more memory but they can be accelerated better with GPU.

**Comment by kisvegabor** [comment]
Something just came to my mind. We can mix to the two ideas (ARGB8888 everywhere and `LV_COLOR_DEPTH` related color format).

Here are the possible cases again:
- **Display has RGB565 format**: no problem, both GPU and CPU can handle RGB565 images, GPU can convert ARGB8888 on the fly to RGB565, CPU can deal with RGB565+A8, ARGB8565 or whatever.
- **Display needs to have transparency** (e.g. LVGL is used to render an OSD menu): TFT driver supports only ARGB8888 for this case (maybe ARGB4444) so no need for supporting ARGB8565 or other similar "hacks". 
- **An LVG layer has RGB565 format**: no problem, same as display with RGB565 format
- **An LVG layer needs to be rendered with alpha** (e.g. a widget with corner radius needs to be drawn on a layer): we can say to support only ARGB8888 in this case. So if a layer needs to have alpha, it's color format will be set the ARGB8888. If it doesn't need to have alpha the native color format will be used (e.g. RGB565).


In summary:
- Use the native format where possible (most of the cases), so we same RAM, flash and memory bandwidth
- For transparent screens support only ARGB8888 (maybe ARGB4444 later to save memory bandwidth)
- For layers with alpha support only ARGB8888. Now there is a dedicated darw_task (see [here](https://github.com/lvgl/lvgl/blob/feat/parallel-rendering/src/draw/sw/lv_draw_sw_img.c#L49-L65) and [here](https://github.com/lvgl/lvgl/blob/feat/parallel-rendering/src/draw/lv_draw.h#L37)) which blends the layers into an other layer (e.g. a rounded rectangle to a the main layer of the screen, maybe also applying transformation), so the whole ARGB8888 to whatever conversation problem is local to the implementation of that draw task. 

What do you think?

**Comment by GorgonMeducer** [comment]
@kisvegabor No problem.

**Comment by kisvegabor** [comment]
To approach it even simpler: 
- we can support any source formats (regardless to the destination format) assuming that there is at least one draw unit that can handle it.
- for output formats we support RGB??? (based on `LV_COLOR_DEPTH`)  and ARGB8888.

And that's it.

If you all agree, I'll update the `feat/parallel-rendering` branch accordingly.

**Comment by tdjastrzebski** [comment]
@kisvegabor this sounds good

**Comment by kisvegabor** [comment]
@nicusorcitu, do you also agree with [this approach](https://github.com/lvgl/lvgl/issues/4059#issuecomment-1481740546)?

**Comment by nicusorcitu** [comment]
Yes, this is fine to me also.

**Comment by kisvegabor** [comment]
I've added the discussed updated. Here is what have been changed:
- These are the native color formats for each possible `LV_COLOR_DEPTH`:
  - 8: L8
  - 16: RGB565
  - 24: RGB888
  - 32: XRGB8888
- So all are without alpha
- If LVGL needs to render with alpha it will always use ARGB8888 (we can add an option to support ARGB4444 later). So when a layer with alpha is required an ARGB8888 layer will be created which later needs to be merged to an other layer with either ARGB8888 or native color format.  
- As dealing with ARGB8888 source and destination buffers will be common they need to optimized in SW image draw functions. (native to ARGB8888, ARGB8888 to native, ARGB8888 to ARGB8888)
- I removed the ARGB8565 color format. As we need RGB565 + alpha only for images RGB565+A8 is enough. In summary these images formats can be used with each color depth if alpha is required too:
  - 8: A8  or L8A8 
  - 16: RGB565+A8
  - 24: ARGB8888
  - 32: ARGB8888

**Comment by GorgonMeducer** [comment]
This is elegant.

**Comment by tdjastrzebski** [comment]
@GorgonMeducer yep, elegant is the right word :)

**Comment by kisvegabor** [comment]
Great! ^^

**Comment by kisvegabor** [comment]
I've pushed a larger update and now most of the widgets work. 

This is what is still missing:
- line and shadow drawing
- `lv_color32_t` on APIs and alpha gradient
- remove all sw masks from the widget level
- image drawing with bitmap mask and/or parametric mask (details below) 
- animimg, bar, canvas, line, slider, colorwheel, chart is still missing
- clip corner
- support stride
- make all demos work
- make all tests pass

All seems clear but "image drawing with bitmap mask and parametric mask" should be discussed. My view about masking is that draw tasks can't receive mask parameters to keep things simple. These are the cases I considered:
- Label, line, and arc masking: not common. 
- A8 masked plain color: consider the A8 mask as an image and draw it with the given color
- Rounded corner mask on a plain color: draw a rounded rectangle

I think we can support A8 mask and rounded rectangle masks only on the images. It has the following advantages: 
- If label, line, or arc needs to be masked (rarely) LVGL can draw them to a layer (image) which can be masked. 
- With clip corner enabled LVGL can draw the children on the corners to a layer and the image can be masked during drawing.
- Text masking from images can be easily supported too as in v9 the font engines always return A8 bitmaps. So maybe we can add support for text masks too. 

In the end IMO it' will have good performance as most to things can be drawn "natively" and masking will be required only once in the last step.

Just to reiterate about the tricky masking cases and their planned solutions:
- Long tick lines on meter: increase the precision of line drawing so short lines can be drawn
- Gradient mask on bar: recalculate the gradient during drawing
- Chart faded area: use right angle triangles and rectangles below them  

It will be different topic these needs to be optimized later:
- dispatching: find independent areas
- break large areas into smaller once (needs to be discusses and benchmarked)
- non masked SW image drawing: 
   - RGB image to RGB layer (memcpy)
   - RGB image to ARGB8888 layer (color conversation might be needed)
   - ARGB8888 image to RGB layer (color conversation and masking)
   - ARGB8888 image to ARGB8888 layer (complex alpha calculation)  
- same masked SW image drawing

**Comment by kisvegabor** [comment]
I've added `lv_color32_t` to style and other APIs but I don't like it. My main problem is that now RGB and A are bound together and it's difficult to set RGB and A separately. For example I want to set only the border opacity to 50% regardless to the actual color. But if we store the color together with the opacity value, we can't change only the opacity (as a value needs to written for the RGB part too in `lv_color32_t`). 

So I think having the color and opa separately  is more flexible. 

To support alpha gradients all we need to do is adding an `lv_opa_t opa` value [here](https://github.com/lvgl/lvgl/blob/38a9c2a8e2cafb3ff3960c7d8a1804ca0b6fd7f7/src/misc/lv_style.h#L152).

What do you think?

**Comment by GorgonMeducer** [comment]
Sorry, I didn't check the original definition of `lv_color32_t`, and can define `lv_color32_t` as follows:
```c
typedef union lv_color32_t {
    struct {
        uint8_t chn[3];
        lv_opa_t opa;
    };
    uint32_t val;
} lv_color32_t;
```

or maybe 
```c
typedef union lv_ccca8888_t {
    struct {
        uint8_t chn[3];
        lv_opa_t opa;
    };
    uint32_t val;
} lv_ccca8888_t;

typedef union lv_accc8888_t {
    struct {
        lv_opa_t opa;
        uint8_t chn[3];
    };
    uint32_t val;
} lv_accc8888_t;
```

**Comment by kisvegabor** [comment]
I don't see how it helps. 

So problem is this:
1. I set the border color in the default state to 0xFFAAAAAA (0xFF is the alpha)
2. I'd like to change ONLY the border alpha in the presses state, so I need to set 0x80AAAAAA

The problem is that in 2) when I change the alpha the RGB will be added too. And RGB part will determine a color,  but I wanted to change only the alpha. If in a 3rd step someone changes the default border color to  0xFF111111, the border color in pressed will be still 0x80AAAAAA.

This way styles, events, or animations cannot change only the alpha as they will set the RGB too.

**Comment by GorgonMeducer** [comment]
> I change the alpha the RGB will be added too...

I don't understand. What does this "added" mean?

**Comment by kisvegabor** [comment]
Let me phrase it differently. This is what happens step by step:
1. I want to set only the border alpha in pressed state. It's not possible directly because, so a new 32 bit color needs to used. E.g. `lv_obj_set_style_border_color(obj, lv_color32(0xffaabbcc), LV_STATE_PRESSED)` (0xff is the alpha). We could read the color of default state first, change the alpha in it, and use that color, but in the end a full 32 bit value will needs to be saved (RGB and A) 
2. During rendering LVGL reads the style properties from the pressed state and it finds the border color (with RGB + A) and it will use it. 

It worked this way. But what if:
1. I set the border color in pressed state as before. (Get the default color, modify the alpha, and use it)
2. An event, other style, or something changes the border color in default state.
3. When LVGL reads the border color in pressed state it will see the RGB + A values. RGB still contains the color from step 1 (before the modification in step 2). It's a problem as I wanted to change only the border's opacity.

The problem is that you can't change only the opacity or the color as they are stored together.

**Comment by GorgonMeducer** [comment]
> you can't change only the opacity or the color as they are stored together.

I don't understand why you cannot change only the opacity.

```c
typedef union lv_color32_t {
    struct {
        uint8_t chn[3];
        lv_opa_t opa;
    };
    uint32_t val;
} lv_color32_t;
```
You can modify the opacity by using `color.opa = xxxx` without modify the RGB part. 
Based on your description, I guess what I said didn't ease your concern. 
But the fact is, I still don't see what the problem is.

I am really sorry. Since I don't understand the problem, I am not able to provide an helpful answer.

**Comment by kisvegabor** [comment]
I will try my best to explain it because I really would like to discuss this topic :slightly_smiling_face: 

In v8, the border color is composed from parts: the RGB part and the Opacity part. They are independent. Therefore it's possible to overwrite only the opacity of the border in pressed state. 

You can modify the opacity in ARGB888, it's fine. The problem is that if we store the RGB and the A part together and we want to set ONLY a different opacity value, you need store the RGB part too, as RGB+A are stored together.  

And once the RGB part is there as well, it will also affect the border color. But I wanted to set only the alpha.

In other words, in v8 if a style stores only `border_opa` we know that the opacity of the border needs to be set accordingly. However with ARGB8888 colors if we want to set only the border opacity we need to set `border_color = 0xffaabbcc`. How will LVGL know that it should ignore the `aabbcc` part and use on the `ff` part for the opacity?

**Comment by GorgonMeducer** [comment]
> How will LVGL know that it should ignore the aabbcc part and use on the ff part for the opacity?

> How will LVGL know that it should ignore the aabbcc part and use on the ff part for the opacity?

How about considering RGB and OPA are CCC + OPA in the same data structure? Or why cannot LVGL just use `ff` part for opacity? What stop you to do that? 

On they other hand, how about add an extra opa for the whole border? So when dealing with ARGB8888, this opa is also calculated together with alpha part of each pixel.

**Comment by kisvegabor** [comment]
> Or why cannot LVGL just use ff part for opacity? What stop you to do that?

In the current implementation of the style system, if a style property is in a style, LVGL knows that the given property needs to be drawn with that value. If a style property is not in a style with a given state the property will be searched with the default state too, or it will be inherited from the parent, or a default value will be used.

When LVGL sees an ARGB8888 color for the pressed border color, like 0x88000000 (0x80 is the alpha, 0x000000 is the RGB part), how it knows that it needs to use only the alpha and ignore the colors.  Note that, I wanted to set only the alpha. But once the whole RGB+A is there, we don't if only the alpha, only the RGB part or both needs to be used.

So the question is how to overwrite only the opacity of the border in pressed state and keep the RGB part from the default state.
In v8:
```c
lv_obj_set_style_border_color(btn, lv_color_hex(0x112233), LV_STATE_DEFAULT); 

//Overwrite only the alpha
lv_obj_set_style_border_opa(btn, 0x88, LV_STATE_PRESSED);

//Change the color only. No problem, it doesn't affect the opacity.  
//In both default and pressed state the color will be 0xAABBCC but with different alpha
lv_obj_set_style_border_color(btn, lv_color_hex(0xAABBCC), LV_STATE_DEFAULT); 
```

With ARGB8888 colors
```c
lv_obj_set_style_border_color(btn, color32(0xff112233), LV_STATE_DEFAULT); 

//I want change only the alpha, so use the color from the default state with different alpha. 
//If the default color changes, it should be used in the pressed state too, just with different alpha
lv_obj_set_style_border_color(btn, color32(0x88112233), LV_STATE_PRESSED);

//Change the color in the default state. Upps, now the pressed and default color doesn't match
lv_obj_set_style_border_color(btn, color32(0xffAABBCC), LV_STATE_DEFAULT); 
```

**Comment by nicusorcitu** [comment]
It means that when you change the border color again, you need to take account of the existing alpha and set:
```lv_obj_set_style_border_color(btn, color32(0x88AABBCC), LV_STATE_DEFAULT);```

Edit: oh, it will change the default one to be the same as the pressed state.

**Comment by nicusorcitu** [comment]
The problem is you want to have same color and two different alpha (for default and pressed state) and clearly that is not possible with a single argb value.

**Comment by kisvegabor** [comment]
Yes, that's it. That's why i suggest keeping the current native color (e.g. RGB565) and separate opacity API and just add an opa component to the gradient descriptors:  
```c
typedef struct {
    lv_color_t color;   /**< The stop color */
    lv_opa_t opa;    // ADD THIS
    uint8_t    frac;    /**< The stop position in 1/255 unit */
} lv_gradient_stop_t;
```

**Comment by nicusorcitu** [comment]
But why on gradient structure? Does every border have a gradient struct? Is this happening only on borders or are other cases too?

**Comment by kdschlosser** [comment]
I do not understand why the opa would not be able to get changed separately if it is attached to the color? if you have a single 32 bit union for the color and only functions to set the color and not the opacity you set up the function so a pointer to the color is what gets passed. 

I am not proficient in C so I am sure the example below is not correct in syntax. But it shows the way to do it.

```c

lv_color_t* color = {.r=255, 'g=255, 'b=255, .a=255};

lv_obj_set_style_bg_color(obj, color);

color->a = 50;
lv_obj_invalidate(obj);
```

The ability to change the opacity is able to be done. It's not done through the use of a function but it still is able to be done,

The side benefit of setting it up this way is you gain the ability to do color comparisons.

```
lv_color_t c1 = lv_color_hex(0x555555);
uint32_t c2 = 0x555555;

if (c2 == (uint32_t) c1) {}
```

The above code will all evaluate to `false` 100% of the time. 



```

lv_obj_set_style_bg_color(obj, lv_color_hex(0x555555), 0);
lv_obj_set_style_bg_opa(obj, 0, 0);

lv_color_t c1 = lv_get_style_bg_color(obj, 0);
uint32_t c2 = 0x00555555;

if (c2 == (uint32_t) c1) {}
```

This also evaluates to `false`

```c
 lv_color_hex(0xFF555555)
```

That just plain fails all together. 

The single largest upside of using the 32 bit union all the time is the ability to set the color depth at runtime. Things like lv_canvas will now work properly. a single function can be made that would handle the color conversions when the buffer gets written to. So the only real expense is the additional 2 bytes used for each of the colors the widgets store. This could be reduced by use of styles so only a single set of colors is stored and they are shared across all widgets. There is no real large expense of doing the conversions at runtime, after all it's only bit shifting being done and processors are very efficient at such a simple math operation. Just like the flush callback function works a function can be made that would handle the bulk writing to the buffer. areas of the screen that are the same color get written to by this bulk writer so a color conversion is only done a single time for an area. calculate the bounding box 3where the color is all the same pass that information along with the color and the buffer to the function and let the function do the conversion and write the data.

IDK if the total program size would end up smaller because of the removal of the get/set opa functions and any other code that handles the opa.

You actually only have a cost of a single byte of memory not 2. I made a boo boo in that statement. so 1 byte of additional memory per set color. on a 32 bit screen you actually have a memory savings not a cost. because you have the 32bit color and then an additional byte for the opa. this way the program would have a smaller footprint. 

It also makes it easier for the user to deal with. the idea of 16 bit color, 24 bit color and 32 bit color causes confusion. it should be a transparent thing for the user. 

On a side note: It would make it possible to supply binaries for the MicroPython binding if the color depth and byte order was able to be set at runtime.

I am working on a CPython binding because of this little guy.

https://onion.io/omega2/

which runs Open WRT and is able to run CPython.

Not all displays that would get connected would use SDL2 and that is a fantastic use case where it would not. So having the ability to set the color depth and byte order at runtime would be beneficial

**Comment by kdschlosser** [comment]
so right now in LVGL it check all of the different states for an object to see if the color matches and if it does it keeps a record of what states match each other and only stores a single color for multiple states?

sounds like a lot of overhead and memory use to do that. first you have to have a container that would be able to store all of the individual colors and opa's along with the states. then checking needs to be done each time a color gets changed.

You also have a waste of memory if the display is a 32 bit display. now you are actually using 40 bits to store a 32 bit color. colors should not be shared unless using a style and the share is only being done between widgets for the exact same attribute. so borders across all widgets, or shadows across all widgets. that is only if the same style is shared between widgets. If a color is set specifically on a widget and a style has been applied then the color gets stored and that gets used instead of the style. So out of the gate when LVGL starts up all widgets should have a style set to it and that style is shared across all widgets by default. This style should be able to be gotten by the user and the style updated if the user wants to make a change to the color across all widgets.

color propagation should take place when an object has a child created for it. When the first child gets made and has a parent the parent would check and see if there are any custom colors that have been set and if there has been a new style gets created and shared between the parent and the children of that parent. This propagation would NOT take place if the child was created without a parent and the child had it's parent set after it's creation. 

If the child's parent changes then the parents style and the child's style would get decoupled from each. colors that are the same as the new parent would be shared with the new parent via a style and all other colors get stored as custom colors. If the style that is set to the child is the same style as the one that is set to the new parent then nothing needs to be done

**Comment by kisvegabor** [comment]
LVGL's style system was inspired by CSS. You can read about the styles in LVGL in more detail [here](https://docs.lvgl.io/master/overview/style.html). 

I think [this comment](https://github.com/lvgl/lvgl/issues/4059#issuecomment-1523332411) summarizes the problem the best way. We can change the alpha channel in ARGB8888 separately. It's fine. The problem is that we can not store only the Alpha value in a style to tell LVGL  to change only the alpha value e.g. in pressed state. If we store the the alpha as part of an ARGB8888 color the RGB part will be there too, and it will overwrite the color in pressed state (not only the alpha).

**Comment by GorgonMeducer** [comment]
As I guessed, others also don't fully understand the issue.
And I now finally see the problem too. My suggestion is to add a dedicated `opa` property. Rather than overriding the alpha channel in each pixel, this one computes with the alpha channel.

**Comment by kisvegabor** [comment]
So do you agree to keep it as it is? :slightly_smiling_face:

**Comment by kdschlosser** [comment]
I understand what is being done. what I do not understand is why it is a problem to change it so that you would set a different color for a different selector. Technically speaking the 4th color channel is just that it IS part of the color and it should be treated if it is a different color, which it is.
so if you have the color 0xFFFFFFFF and you have the color 0x55FFFFFF are they different colors? the answer is yes. it is no different then 0xFFFFFF and 0x55FFFF being different colors because they are. 0xFFFFFFFF and 0xFFFFFF are not the same color. they are visually the same but they are not the same. if you place 0x55FFFFFF on top of a background that is 0xFFFFFFFF it is going to visually appear to be the same but once again it is in fact a completely different color.


You still also have the fact of the extra memory use if the color depth is set to 32 bit because the alpha channel is not being used in lv_color_t, so for every single color that is set there is a byte that is wasted. How often does just the opacity change when the state of a widget changes. more then likely on when wantiong to turn off the displaying of that element. If the hidden flag is set I would imagine that the rendering of the widget is completely bypassed and not the opacity is changed to zero. how come turning on and off elements in the widget is only able to be do by color manipulation?

Now I do have a question. when I set the border of a widget to have an opacity of zero does LVGL still render the border?. If LVGL still renders it even tho the opacity is set to zero that would not be ideal from a resources aspect.

I suggest using flags instead of the opa to turn on and off the displaying of various elements of a widget. it could be added into the same spot the state flags are stored if there is room to do so. a simple bit wise check of the set flags would determine if the rendering of that specific component of the widget should be rendered and if no then completely bypass it. 

This would be a better solution in all reality because of  "set it and forget it". so the color with opacity could be set and it being displayed could easily be turned on and off without the user having to explicitly code in what the opacity level of the ops is to  turn it back on. Or the using having to code in a place to store the opa level before setting it to zero so it can be turned back on to the correct level.

**Comment by kisvegabor** [comment]
> Now I do have a question. when I set the border of a widget to have an opacity of zero does LVGL still render the border?

No, it's not drawn. See [here](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw_rect.c#L397-L400).

> I suggest using flags instead of the opa to turn on and off the displaying of various elements of a widget. it could be added into the same spot the state flags are stored if there is room to do so. a simple bit wise check of the set flags would determine if the rendering of that specific component of the widget should be rendered and if no then completely bypass it.

It really not just an on/off thing and it applies to all colors. (bg, border, outline, shadow, arc, text, etc). Adding such a limitation would be a larger problem than losing a few bytes in case of some styles.

**Comment by kdschlosser** [comment]
> It really not just an on/off thing and it applies to all colors. (bg, border, outline, shadow, arc, text, etc). Adding such a limitation would be a larger problem than losing a few bytes in case of some styles.

I guess my question is what exactly is the problem? I have the ability to say set the color of a background on a widget in the default state and then set a color to the background for the pressed state. It works like it should work. I am imagining the code for the pressed state checking to see if a color has been set and if it is NULL then it uses the default color instead. the memory is still allocated even if the color for the pressed state has not been filled in. The change that would need to be made is instead of it being NULL it would be populated with the default color upon creation.

I am not understanding how moving where the alpha/opa value is stored would in any way limit the functionality.

**Comment by kisvegabor** [comment]
>  the memory is still allocated even if the color for the pressed state has not been filled in.

Nope, the memory is not allocated. We store only those properties which are really added. If a property is not found (i.e. not added to any related styles), we consider it unset, and try inheriting it, or use the default value, etc.

**Comment by kdschlosser** [comment]
I still don't understand where there is a limitation if a 32bit color was stored as a 32 bit color not as a 24 bit color and an 8 bit alpha channel. The other thing is what is the purpose of the 32 bit color when it is only used for images and gradients. Those are the only things that have the ability to use the alpha channel in the 32 bit color. so that extra byte goes unused for a large portion of the time. may as well put the color in a structure that has 3 8 bit fields so the extra byte that goes unused doesn't take up memory.

right now when LVGL handles rendering the colors if there is an alpha channel being used that is less than 255 and the bit depth is set to 16 the colors produced by LVGL can be incorrect. if you convert 2 colors to 16 bit and then lay those colors on top of one another having the alpha channel set to say 80 there is the possibility of the color not being correct. if you mix 2 24 bit colors (technically 32 bit because of the alpha channel) and then convert the final color to 16 bit it is not always going to be the same color as what is gotten if the 2 colors are converted to 16 bit before they are mixed. so by using RGB888 all the time and only doing the conversion to RGB565 as the very last thing to do the colors are going to end up being correct.  how would a 24 bit color gets stored without taking up the space of a 32 bit color?

**Comment by kisvegabor** [comment]
To make searching faster all style properties are stored in a 32 bit variables (64 bit on 64 bit systems). Maybe we can optimize it, but it's a different topic and we should discuss it separately. 

Let's say we have a RGB565 display so we need RGB565 result somehow. When we need to fill it, it's fine, just set the pixels to the given color. When we need to draw a simple RGB565 image on it, it's also fine, just a 2D memcpy. 
However, we need to really mix RGB565 (with some oapcity), RGB565+A8 or ARGB8888. So the target buffer always has RGB565 format in these case the the source data can have multiple formats. Why don't we always use (A)RGB(8)888 as target buffer buffer and convert it to RGB565 in the end? 
1. to save memory
2. to save the time of conversation

Anyway, as it's very important to support setting the opacity independently from the RGB part, we have no option, we need 2 properties for each color related style properties: RGB and opacity.

**Comment by GorgonMeducer** [comment]
> Anyway, as it's very important to support setting the opacity independently from the RGB part, we have no option, we need 2 properties for each color related style properties: RGB and opacity.

Agree with you.

**Comment by nicusorcitu** [comment]
Hi Gabor,

I am trying to add the support for LV_DRAW_TASK_TYPE_IMAGE on GPU but there are too many corner cases.
I think we need a proper document to explain what exactly shall cover each task. For this particular TASK_IMAGE we need to know:
- what kind of formats shall it cover?
- w/ transformation: rotate and zoom (does it have to cover also rotate + zoom at the same time?)
- w/ recolor (does it have to support recolor + zoom or recolor + rotate or recolor + zoom + rotate at the same time?)
- w/ or w/o mask?
- color key?

Most important thing: you mentioned that GPU shall never fallback to software.

But what if a GPU or DMA does not support a task with all combinations above? What shall it do?
1. From GPU routine we need to call the software routine and handle the task.
E.G. If the rotation angle is 45 and it support only 90 degree, shall i have?

```
if(draw_dsc->angle != 900) {
    lv_draw_sw_img((lv_draw_unit_t *)u, t->draw_dsc, &t->area);
    return;
}

```
2. Return invalid and mark that task as LV_DRAW_TASK_STATE_QUEUED so that the SW thread can pick it up.
```
if(draw_dsc->angle != 900) {
    return -1;
}

```
Then the GPU thread callback will read the status of the `gpu_execute_drawing` and if -1 it will set the state back

        /*Set the task state back to queued so that SW can take it. But how can we know that GPU dispatcher is not going to pick exactly this one again?*/
        u->task_act->state = LV_DRAW_TASK_STATE_QUEUED ;
        u->task_act = NULL;

        /*The draw unit is free now. Request a new dispatching as it can get a new task*/
        lv_draw_dispatch_request();

Thanks,
Nicu.

**Comment by kisvegabor** [comment]
> what kind of formats shall it cover?

If you mean color formats, than all the formats that the GPU can support. However note that the target color format is always RGB565/888 or ARGB8888. 

> w/ transformation: rotate and zoom (does it have to cover also rotate + zoom at the same time?)

Yes, it should cover it if it can support it.

> w/ recolor (does it have to support recolor + zoom or recolor + rotate or recolor + zoom + rotate at the same time?)

Yes, if it can do it with a reasonable performance.

> w/ or w/o mask?

Masks shouldn't be parts of the images, I've just removed it. There is a dedicated `LV_DRAW_TASK_TYPE_MASK_RECTANGLE` which should mask out a rounded rectangle from an ARGB8888 image. We can add more mask related draw tasks later but only it was required from me now.

> color key?

Yes, if possible.


> From GPU routine we need to call the software routine and handle the task.
E.G. If the rotation angle is 45 and it support only 90 degree, shall i have?

The GPU shouldn't take a draw task if it can't draw it. If a draw task is not good for the GPU just call  `lv_draw_get_next_available_task` again to see if there is a good draw task.

> Return invalid and mark that task as LV_DRAW_TASK_STATE_QUEUED so that the SW thread can pick it up.

If the were candidates but the GPU couldn't pick any tasks, just `return 0`. `return -1` means there are no free draw tasks at all, so other dispatchers shouldn't be called.

---

I think this concept is quite abstracts and flexible, but if you think it is not working well, just let me know and we can figure out something. 

Note that you can say that a rotated+zoomed+color keyed+recolor image in ARGB2222 color format could be handled by the GPU but it's too corner case. So just don't take it and leave it for SW. It's ok to support only the simpler combinations as they will cover most of the use cases.

**Comment by nicusorcitu** [comment]
> If you mean color formats, than all the formats that the GPU can support. However note that the target color format is always RGB565/888 or ARGB8888.
>
I still see code with LV_COLOR_FORMAT_RGB565A8.
The layer blending is also covered by this task at the end. What is the layer format? NATIVE if there is no alpha and always ARGB8888 with alpha?

> Masks shouldn't be parts of the images, I've just removed it. There is a dedicated LV_DRAW_TASK_TYPE_MASK_RECTANGLE which should mask out a rounded rectangle from an ARGB8888 image. We can add more mask related draw tasks later but only it was required from me now.
>
1. Then why do I see this code in `lv_draw_sw_img`?
```
    bool masked = (draw_dsc->bitmap_mask && _lv_area_is_on(&draw_dsc->bitmap_mask_area, &transformed_area)) ||
                  _lv_area_is_on(&draw_dsc->rectangle_mask_area, &transformed_area);
```
2. You created a separate task just for rounding rectangles? I thought the masking widgets will get draw on a layer an that layer will be blended with a A8 mask as a final operation.

> The GPU shouldn't take a draw task if it can't draw it. If a draw task is not good for the GPU just call lv_draw_get_next_available_task again to see if there is a good draw task.
>
So you are suggesting the `lv_draw_gpu_dispatch` shall do all these checks for a particular task. E.G. if the gpu support rotation or zoom but not both at the same time, this case shall be discovered on dispatcher (in a switch case statement for every task which is ready to draw). If the "supported" checks are failing, that task shall be skiped.
But if the checks are passing, almost the same verification will be redone while executing the task. Almost duplicating the code.

**Comment by kisvegabor** [comment]
> What is the layer format? NATIVE if there is no alpha and always ARGB8888 with alpha?

Exactly.

> Then why do I see this code in lv_draw_sw_img?

I've removed it yesterday.

> So you are suggesting the lv_draw_gpu_dispatch shall do all these checks for a particular task. E.G. if the gpu support rotation or zoom but not both at the same time, this case shall be discovered on dispatcher.

Yes. However, it can be possible to configure the GPU to take these tasks too and execute the rotation+zoom in 2 steps, assuming that it's faster/better then SW.

> But if the checks are passing, almost the same verification will be redone while executing the task. Almost duplicating the code.

Do you mean it's redundant to add the same bunch of `if`s? If so yes, it is... But I think it's the smaller problem.

**Comment by kisvegabor** [comment]
Not stale.

**Comment by kdschlosser** [comment]
not stale

**Comment by kdschlosser** [comment]
not stale

**Comment by kisvegabor** [comment]
Status update:
Most of the things seems working in `feat_parallel-rendering`. I've also added a bunch of rendering tests.

There are a few things that are not reimplemented yet and should be discussed:
- Do we really need chroma keyed image support? In the practice I never use it, and I think it isn't worth the extra complication. 
- Same for sub pixel rendered text. It's very tricky to blend because it has  separate alpha channels for red, green, and blue. In SW render I needed to implement a blender for each target color formats (RGB565, RGB888, XRGB888, ARGB888). I think GPUs don't support it either so I would drop it as well.
- Image recoloring: I'm still thinking about handling with layers. So create a layer, have an image recolor draw task, and blend the recolored image. The other option is to recolor on the fly while blending (as now). Could you comment on how it works with the GPUs?

cc @nicusorcitu @GorgonMeducer

**Comment by kdschlosser** [comment]
The recoloring of images is something I do use. It's an easy way to handle changing the color of a gradient. I make an image of say a radial gradient and I make it from white to alpha and then I can place that gradient as a child of an object. Now I have the ability to set 2 colors for the gradient. One being the parent objects background color and the second is doing an image recolor of the gradient.

**Comment by GorgonMeducer** [comment]
> Do we really need chroma keyed image support? In the practice I never use it, and I think it isn't worth the extra complication.

Unless you want to support 8-bit game design, it is almost no one uses the chrome-keyed image in practice. The only exception is transform, with a chrome-keyed image, you can ignore many pixels and hence improve performance.

**Comment by GorgonMeducer** [comment]
> Same for sub pixel rendered text. It's very tricky to blend because it has separate alpha channels for red, green, and blue. In SW render I needed to implement a blender for each target color formats (RGB565, RGB888, XRGB888, ARGB888). I think GPUs don't support it either so I would drop it as well.

I haven't looked at the sub-pixel, and I don't understand its meaning. Sorry. But it looks like you can use ARGB8888 as the internal format. Arm-2D can support individual colour channel access for all 32-bit colour formats; maybe other GPUs can do similar work.

But since I haven't implemented the arm-2d acceleration for the sub-pixel, I am OK with removing this feature.

**Comment by GorgonMeducer** [comment]
> Image recoloring: I'm still thinking about handling with layers. So create a layer, have an image recolor draw task, and blend the recolored image. The other option is to recolor on the fly while blending (as now). Could you comment on how it works with the GPUs?

This is how it is accelerated in arm-2d:

https://github.com/lvgl/lvgl/blob/65f031c2bd33bc4b5c135c1c41213ee25c644000/src/draw/arm2d/lv_gpu_arm2d.c#L281-L316

**Comment by kisvegabor** [comment]
@kdschlosser 
> The recoloring of images is something I do use. It's an easy way to handle changing the color of a gradient. 

Hm, it's a good idea! I've never thought of that.

@GorgonMeducer 
> This is how it is accelerated in arm-2d:

Won't is overwrite the alpha channel of original image e.g. in case of ARGB8888 format?

**Comment by nicusorcitu** [comment]
1. I do not know exactly the scenario of using the chroma key in MCU, maybe the QR code scanner might use this feature.
But I would not remove the chroma key, it has been shipped as feature for so long and probably someone is already using it.
PXP support chroma key and also the new VGLite GPU will have this support. So if possible, please leave it in place on lvgl v9.

3. PXP currently support image recoloring. But VGLite GPU will support it too - just need to add the support on lvgl.
PXP have two surface processing, Alpha Surface (AS) and Process Surface (PS) which are blend based on Porter Duff configuration.
For recoloring, we put the source image in AS and disable PS (only set a register in PS for recoloring). The recolor is done on the fly, just by fetching the AS.
If you want to create an layer and use it for recoloring, that layer will need to be put in PS and blend with AS. This means that PS layer will need to be fetched too for blending, so this will be an extra time.

**Comment by kisvegabor** [comment]
**Chroma key** 
As v9 will be major rework anyway, so IMO updating a few images shouldn't be a big issue. To prevent bloating LVGL  I wouldn't like to keep it just because it's doable. 

**Image recolor** 
I've implemented on the fly (without layers) recoloring in v9. It works the same as in v8. See the tests [here](https://github.com/lvgl/lvgl/blob/feat/parallel-rendering/tests/ref_imgs/draw/image_format_recolor.png) and [here](https://github.com/lvgl/lvgl/blob/feat/parallel-rendering/tests/ref_imgs/draw/image_format_rotate_and_recolor.png).

**Sub px rendering**  
If there is no objection I'd remove it for now.

**Comment by GorgonMeducer** [comment]
> Won't is overwrite the alpha channel of original image e.g. in case of ARGB8888 format?

It won't. It is equivalent to the sw implementation.

**Comment by kisvegabor** [comment]
All v8 features are working again and `feat/parallel-rendering` is merged to `master` so I close this issue.

**Comment by W-Mai** [comment]
> All v8 features are working again and feat/parallel-rendering is merged to master so I close this issue.

I notice that clip_corner for lv_img doesn't take effect.

Hasn't the masking been fully realized?

**Comment by W-Mai** [comment]
> I think what you mean is if an image has a radius and clip corner is enabled you would like to clip the image corners without a wrapper item, right?

That's exactly what I meant!

**Comment by kisvegabor** [comment]
For the sake of consistence , would the the wrapper object approach work in your case?

**Comment by W-Mai** [comment]
> For the sake of consistency , would the the wrapper object approach work in your case?

<img width="226" alt="image" src="https://github.com/lvgl/lvgl/assets/24841247/33eea0e6-2c7f-4cfd-a365-874804e97cbf">

the wrapper object works correctly

**Comment by kisvegabor** [comment]
Cool!
